designs/node-ownership.md

```diff
@@ -0,0 +1,227 @@
+
+The recommendation is that this in-flight capacity should be modeled as an internal Machine CR that represents a Karpenter provisioner/consolidation scheduling decision. This has the following benefits:
+
+1. Karpenter stops creating the Node object, avoiding race conditions and conflicts with the kube-apiserver and the kubelet that are causing orphaned nodes and instances
+2. Karpenter does not need to know the node name at the end of its provisioning loop, shortening the provisioning loop for clouds that do not know the node name at instance launch time
+3. Karpenter stops mutating the Node object by adding labels, ownerReferences and finalizers, avoiding conflicts with other Kubernetes tooling which relies on a well-known state for the Node.
+4. Karpenter stops finalizing nodes such that nodes hang on deletion when Karpenter is uninstalled.
+5. Karpenter makes its in-flight capacity observable to users through well-known Kubernetes APIs
+
+_Note: This internal Machine CR will come in as an alpha API and should __not__ be relied upon for any dependent tooling until the API is stable._
+
+## Background
+
+Karpenter currently creates the node object on the Kubernetes api server immediately after creating the VM instance. Kubernetes cloud providers (EKS, AKS, GKE, etc.) assume that, ultimately, the kubelet will be the entity responsible for registering the node to the api-server. This is reflected [through the userData](https://github.com/awslabs/amazon-eks-ami/blob/master/files/bootstrap.sh) where KubeletConfig can be set [that is only properly propogated for all values when the kubelet is the node creator](https://github.com/kubernetes/kubernetes/blob/39c76ba2edeadb84a115cc3fbd9204a2177f1c28/pkg/kubelet/kubelet_node_status.go#L286). However, Karpenter’s current architecture necessitates that it both launches the VM instance and creates the node object on the Kubernetes API server in succession (more on this [below](#why-does-karpenter-createoperate-on-the-node-at-all)).
+
+This document describes the current node creation flow for Karpenter as well as the rationale for why Karpenter originally created the node object. It then calls out the specific problems with this approach and recommends an alternative approach to creating the Node object that solves for the current approach’s problems.
+
+### Current Instance & Node Creation Flow
+
+Karpenter currently performs the following operations when it detect pending pods and launches capacity to the cluster:
+
+1. Performs a **CreateFleet** request against EC2 APIs which leads to VM Instance Creation
```

Comment:
Would that really be a MachineSet rather than a Machine (both could still be custom resources that Karpenter defines)?

Reply:
Can you describe what you mean? I'm not seeing how the line you are pointing at here links to the question?

Reply:
Ah, what I meant is that `CreateFleet` creates a _set_ of machines, so the CR should be a MachineSet not a Machine.

Maybe there's a good explanation why this is wrong; for example, we tell people writing cloud provider integrations that Karpenter must only ever create fleets (and similar abstractions) with a max of 1 running server.

Reply:
In our case, when we make a `CreateFleet` call, we aren't asking for a set of machines, we're really asking for a single machine at a time, which we batch together if the requirements are "similar enough"

It also makes the management experience post-launch difficult to reason about since there isn't a way to represent each Machine as a single entity

Reply:
We _could_ model this as a MachineSet _and_ a set of Machines, but it sounds like there's little need.